### PR TITLE
Ospi with DMA transfer for stm32l5x and stm32H7 devices

### PR DESCRIPTION
**Pull request by FRASTM**
this PR enables the octoSPI NOR flash that is present on the
-   stm32l562e_dk disco kit 
- stm32h735g_disco and stm32b3i_dk disco kits

The octospi peripheral are defined by the DTS of the devices.
The boards enable the mx25lm51245 NOR octo-flash controller

The sdfp table is included as a DTS property because NOR flash does not answer correctly.

It requires https://github.com/zephyrproject-rtos/zephyr/pull/39770
Requires the pin definition with https://github.com/zephyrproject-rtos/hal_stm32/pull/122

Signed-off-by: Francois Ramu [francois.ramu@st.com](mailto:francois.ramu@st.com)

### CODE REVIEW

**Comment by erwango** [comment]
^^@nfaurant

**Comment by FRASTM** [comment]
running without DEBUG_CONFIG : when the sdfp table has 9 dwords (32-bit elements).

**Comment by JiiZR** [comment]
I tried this with b_u585i_iot02a kit. Also merged manually file stm32u585aiixq-pinctrl.dts from https://github.com/zephyrproject-rtos/hal_stm32/pull/122. Then I tried to build with command: west build -p auto -b b_u585i_iot02a samples/drivers/spi_flash but there was build problem at file stm32u5xx_hal_ospi.c, I fixed that commenting line
      //status = HAL_DMA_Abort(hospi->hdma);.
I flashed binary and everything seemed to be fine, but then I changed test pattern from 	const uint8_t expected[] = { 0x55, 0xaa, 0x66, 0x99 }; to 	const uint8_t expected[] = { 0x55, 0xaa, 0x66, 0x9f };. After running binary results were:
*** Booting Zephyr OS build zephyr-v3.0.0-3054-g920db04a245c  ***

JEDEC OSPI-NOR SPI flash testing
==========================

Test 1: Flash erase
Flash erase succeeded!

Test 2: Flash write
Attempting to write 4 bytes
Data read does not match data written!!
00010000 wrote 55 read 55 match
00010001 wrote aa read aa match
00010002 wrote 66 read 66 match
00010003 wrote 9f read 99 MISMATCH

I found that flash_erase is not working, I added memory read prints before and after flash_erase. I was wondering if I still miss some code or is this real problem ?

**Comment by FRASTM** [comment]
> I tried this with b_u585i_iot02a kit. Also merged manually file stm32u585aiixq-pinctrl.dts from [zephyrproject-rtos/hal_stm32#122](https://github.com/zephyrproject-rtos/hal_stm32/pull/122). Then I tried to build with command: west build -p auto -b b_u585i_iot02a samples/drivers/spi_flash but there was build problem at file stm32u5xx_hal_ospi.c, I fixed that commenting line //status = HAL_DMA_Abort(hospi->hdma);. I flashed binary and everything seemed to be fine, but then I changed test pattern from const uint8_t expected[] = { 0x55, 0xaa, 0x66, 0x99 }; to const uint8_t expected[] = { 0x55, 0xaa, 0x66, 0x9f };. After running binary results were: *** Booting Zephyr OS build zephyr-v3.0.0-3054-g920db04a245c ***
> 
> # JEDEC OSPI-NOR SPI flash testing
> Test 1: Flash erase Flash erase succeeded!
> 
> Test 2: Flash write Attempting to write 4 bytes Data read does not match data written!! 00010000 wrote 55 read 55 match 00010001 wrote aa read aa match 00010002 wrote 66 read 66 match 00010003 wrote 9f read 99 MISMATCH
> 
> I found that flash_erase is not working, I added memory read prints before and after flash_erase. I was wondering if I still miss some code or is this real problem ?

The fix is coming with the support of DMA with stm32l562

**Comment by FRASTM** [comment]
Preparing the support of the DMA for octo-spi NOR flash on the u585 disco kit : https://github.com/FRASTM/zephyr/tree/ospi_dma

**Comment by FRASTM** [comment]
When DMA transfer is used for the OctoSPI read/write : running on the stm32l562e_dk board: 
 _$ west build -p auto -b stm32l562e_dk samples/drivers/spi_flash_
```
*** Booting Zephyr OS build zephyr-v3.0.0-3068-g98cc3c431b0d  ***

JEDEC OSPI-NOR SPI flash testing
==========================

Test 1: Flash erase
Flash erase succeeded!
                                                              
Test 2: Flash write                                           
Attempting to write 4 bytes                                   
Data read matches data written. Good!!      
```                  
running on the b_u585i_iot02a board: 
 _$ west build -p auto -b b_u585i_iot02a samples/drivers/spi_flash_
```
*** Booting Zephyr OS build zephyr-v3.0.0-3068-g98cc3c431b0d  ***

JEDEC OSPI-NOR SPI flash testing
==========================

Test 1: Flash erase
Flash erase succeeded!
                                                                                  
Test 2: Flash write                                                               
Attempting to write 4 bytes                                                       
Data read does not match data written!!                                           
00010000 wrote 55 read 55 match                                                   
00010001 wrote aa read aa match                                                   
00010002 wrote 66 read 66 match                                                   
00010003 wrote 99 read 99 match   
```

**Comment by FRASTM** [comment]
When DMA is used for octoSPI flash on the stm32u585, the GPDMA1 is configured
and the stm32 dma driver is updated to support that new peripheral
The DTS is set to enable the dma request for the octospi2 : 
```
&octospi2 {
+            /* channel 12-15 are for transfers to/from external memories */
+            dmas = <&gpdma1 12 41 0x10480>; /* request 41 for OCTOSPI2 */
+            dma-names = "tx_rx";
+};
```
According to the stm32u5x refMan:

Running on the b_u585i_iot02a board: 
 _$ west build -p auto -b b_u585i_iot02a samples/drivers/spi_flash_
```
*** Booting Zephyr OS build zephyr-v3.0.0-3068-g98cc3c431b0d  ***

JEDEC OSPI-NOR SPI flash testing
==========================

Test 1: Flash erase
Flash erase succeeded!
                                                                                  
Test 2: Flash write                                                               
Attempting to write 4 bytes                                                       
Data read does not match data written!!                                           
00010000 wrote 55 read 55 match                                                   
00010001 wrote aa read aa match                                                   
00010002 wrote 66 read 66 match                                                   
00010003 wrote 99 read 99 match   
```

**Comment by FRASTM** [comment]
Running on the b_u585i_iot02a board or smt32l562e_dk in **SPI/STR** mode, the 4 written and read data are identical,
but memcmp returns  <>0 though

**Comment by FRASTM** [comment]
rebased after merge of the octospi driver?

File: dts/bindings/ospi/st,stm32-ospi.yaml [hunk 1]
```diff
@@ -33,3 +36,33 @@ properties:
...
+           dmas = <&dma1 5 41 0x10000>;
+
+        With, in each cell of the dmas specificier:
+        - &dma1: dma controller phandle
+        - 5: channel number (0 to Max-Channel minus 1). From 10 to 15 on stm32u5x.
+        - 41: slot number (request which could be given by the DMAMUX)
+        - 0x10000: channel configuration (only for srce/dest data size, priority)
+
+        Notes:
+          - On series supporting DMAMUX, transnational DMA phandle should
```
Comment (erwango) [inline dts/bindings/ospi/st,stm32-ospi.yaml:53]: transnational ?

File: drivers/flash/flash_stm32_ospi.c [hunk 2]
```diff
@@ -37,6 +44,27 @@ LOG_MODULE_REGISTER(flash_stm32_ospi, CONFIG_FLASH_LOG_LEVEL);
 #define STM32_OSPI_SUBSECTOR_4K_ERASE_MAX_TIME  400U
 #define STM32_OSPI_WRITE_REG_MAX_TIME           40U
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dmamux)
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:47]: dma could be enabled for other purpose than ospi.
What about the following:
`#if STM32_OSPI_USE_DMA`

File: dts/bindings/ospi/st,stm32-ospi.yaml (hunk 1 above)
Comment (FRASTM) [reply dts/bindings/ospi/st,stm32-ospi.yaml:53]: fixed

Comment (mbolivar-nordic) [reply dts/bindings/ospi/st,stm32-ospi.yaml:53]: Did you forget to squash or push the fix? I still see it here

File: dts/bindings/ospi/st,stm32-ospi.yaml [hunk 3]
```diff
@@ -33,3 +36,33 @@ properties:
...
     pinctrl-names:
       required: true
+
+    dmas:
+      description: |
+        Optional DMA channel specifier, required for DMA transactions.
+        For example dmas for TX/RX on OSPI
+           dmas = <&dma1 5 41 0x10000>;
+
+        With, in each cell of the dmas specificier:
```
Comment (mbolivar-nordic) [inline dts/bindings/ospi/st,stm32-ospi.yaml:46]: specificier -> specifier

**Review by mbolivar-nordic — COMMENTED** [review]
I think this will have to wait for v3.2, so I will point out another typo in the meantime while it is being reviewed. Otherwise the devicetree changes look fine to me!

File: dts/bindings/ospi/st,stm32-ospi.yaml (hunk 3 above)
Comment (FRASTM) [reply dts/bindings/ospi/st,stm32-ospi.yaml:46]: fixed

File: drivers/flash/flash_stm32_ospi.c [hunk 4]
```diff
@@ -329,7 +368,17 @@ static int stm32_ospi_mem_ready(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uin
...
+		s_command.Instruction = (nor_mode == OSPI_SPI_MODE)
+				? SPI_NOR_CMD_RDSR
+				: SPI_NOR_OCMD_RDSR;
+		s_command.AddressMode = (nor_mode == OSPI_SPI_MODE)
+				? HAL_OSPI_ADDRESS_NONE
+				: HAL_OSPI_ADDRESS_8_LINES;
+		s_command.Address = 0;
+		s_command.DummyCycles = (nor_mode == OSPI_SPI_MODE)
+				? 0U
+				: (nor_rate == OSPI_DTR_TRANSFER)
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:381]: This piece of code is inside if (nor_mode == OSPI_OPI_MODE) {, so result of the initial comparison is known.

File: drivers/flash/flash_stm32_ospi.c [hunk 5]
```diff
@@ -329,7 +368,17 @@ static int stm32_ospi_mem_ready(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uin
 	s_command.Address = 0U;
 
 	if (nor_mode == OSPI_OPI_MODE) {
-		s_command.DummyCycles = (nor_rate == OSPI_DTR_TRANSFER)
+		s_command.DummyCycles = (nor_rate == OSPI_DTR_TRANSFER);
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:371]: `s_command.DummyCycles` is now 0 or 1, is that expected ?

File: drivers/flash/flash_stm32_ospi.c [hunk 6]
```diff
@@ -329,7 +368,17 @@ static int stm32_ospi_mem_ready(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uin
...
+		s_command.Instruction = (nor_mode == OSPI_SPI_MODE)
+				? SPI_NOR_CMD_RDSR
+				: SPI_NOR_OCMD_RDSR;
+		s_command.AddressMode = (nor_mode == OSPI_SPI_MODE)
+				? HAL_OSPI_ADDRESS_NONE
+				: HAL_OSPI_ADDRESS_8_LINES;
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:377]: This piece of code is inside `if (nor_mode == OSPI_OPI_MODE) {`, so result of those comparison is known.

File: drivers/flash/flash_stm32_ospi.c (hunk 4 above)
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:381]: Question on the whole section: Is that related to DMA? Or hould it be moved in a dedicated commit ?

File: drivers/flash/flash_stm32_ospi.c [hunk 7]
```diff
@@ -79,6 +107,10 @@ struct flash_stm32_ospi_data {
 	enum jesd216_mode_type read_mode;
 	enum jesd216_dw15_qer_type qer_type;
 	int cmd_status;
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dmamux)
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:111]: Can't you use `STM32_OSPI_USE_DMA` here ?

File: drivers/flash/flash_stm32_ospi.c [hunk 8]
```diff
@@ -1063,6 +1112,22 @@ __weak HAL_StatusTypeDef HAL_DMA_Abort(DMA_HandleTypeDef *hdma)
 	return HAL_OK;
 }
 
+/* This function is executed in the interrupt context */
+#if STM32_OSPI_USE_DMA
+static void ospi_dma_callback(const struct device *dev, void *arg,
+			 uint32_t channel, int status)
+{
+	DMA_HandleTypeDef *hdma = arg;
+
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:1121]: `ARG_UNUSED(dev);`

File: drivers/flash/flash_stm32_ospi.c [hunk 9]
```diff
@@ -1571,6 +1636,75 @@ static int flash_stm32_ospi_init(const struct device *dev)
...
+	 * HAL expects a valid DMA channel (not DMAMUX).
+	 * The channel is from 0 to 7 because of the STREAM_OFFSET in the dma_stm32 driver
+	 */
+	hdma.Instance = __LL_DMA_GET_CHANNEL_INSTANCE(dev_data->dma.reg,
+						      dev_data->dma.channel);
+#else
+	hdma.Instance = __LL_DMA_GET_CHANNEL_INSTANCE(dev_data->dma.reg,
+						      dev_data->dma.channel-1);
+#endif
+#endif /* CONFIG_DMA_STM32_V1 */
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:1699]: Shouldn't we limit to L5 case only ?
Or are we confident that this code is ok for other cases also ?
(For instance do we have configurations with `CONFIG_DMA_STM32_V1` + OSPI ?)

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by FRASTM** [comment]
it appears that after the merge of https://github.com/zephyrproject-rtos/zephyr/pull/45917 the samples/drivers/spi_flash fails when tested on stm32l562e_dk

**Comment by FRASTM** [comment]
rebase
remove code which is not related to DMA 
add the table_priority as in PR #46770

Note that because of issue https://github.com/zephyrproject-rtos/zephyr/issues/46740, the ospi transfer with DMA fails on the 
stm32l562e_dk board

File: drivers/flash/flash_stm32_ospi.c (hunk 9 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: removed as there is no other mcu than stm32L5 for ospi transfer with dma

File: drivers/flash/flash_stm32_ospi.c (hunk 7 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c:119]: changed to #if STM32_OSPI_USE_DMA

File: drivers/flash/flash_stm32_ospi.c (hunk 4 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: no more applicable

File: drivers/flash/flash_stm32_ospi.c (hunk 6 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: no more applicable

**Comment by FRASTM** [comment]
rebase on f3017aa787e75c77bb7deb9f4ae45d4b4e93fbc3

**Comment by FRASTM** [comment]
Testing  samples/drivers/spi_flash on stm32l562e_dk board:
in OSPI with DMA transfer, for OSPI DTR or STR and SPI STR modes

```
*** Booting Zephyr OS build zephyr-v3.1.0-3687-ga6590ddf88d9  ***

ospi-nor-flash@0 SPI flash testing
==========================

Test 1: Flash erase
Flash erase succeeded!

Test 2: Flash write
Attempting to write 4 bytes
Data read matches data written. Good!!
```

**Review by erwango — DISMISSED** [review]

File: dts/bindings/ospi/st,stm32-ospi.yaml (hunk 3 above)
Comment (GeorgeCGV) [reply dts/bindings/ospi/st,stm32-ospi.yaml:46]: The original error is present.
```suggestion
        With, in each cell of the dmas specifier:
```

**Comment by FRASTM** [comment]
rebase on e5b049249eab5a9bc18a5575fd052e7f7d5bc41e

File: drivers/flash/flash_stm32_ospi.c [hunk 10]
```diff
@@ -1081,6 +1127,25 @@ __weak HAL_StatusTypeDef HAL_DMA_Abort(DMA_HandleTypeDef *hdma)
...
+
+/* This function is executed in the interrupt context */
+#if STM32_OSPI_USE_DMA
+static void ospi_dma_callback(const struct device *dev, void *arg,
+			 uint32_t channel, int status)
+{
+	DMA_HandleTypeDef *hdma = arg;
+
+	ARG_UNUSED(dev);
+
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:1139]: ARG_UNUSED(status);

File: drivers/flash/flash_stm32_ospi.c [hunk 11]
```diff
@@ -1081,6 +1127,25 @@ __weak HAL_StatusTypeDef HAL_DMA_Abort(DMA_HandleTypeDef *hdma)
...
+	if (status != 0) {
+		LOG_ERR("DMA callback error with channel %d.", channel);
+
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:1142]: Remove empty line

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@GeorgeCGV Is that fine for you ?

**Comment by GeorgeCGV** [comment]
@erwango I didn't test DMA and other merged OSPI related changes yet. Looks ok.

**Review by GeorgeCGV — APPROVED** [review]
